datafusion/optimizer/src/type_coercion.rs

```diff
@@ -58,9 +58,19 @@ impl OptimizerRule for TypeCoercion {
     fn optimize(
         &self,
         plan: &LogicalPlan,
-        _optimizer_config: &mut OptimizerConfig,
+        optimizer_config: &mut OptimizerConfig,
     ) -> Result<LogicalPlan> {
-        optimize_internal(&DFSchema::empty(), plan)
+        Ok(self
+            .try_optimize(plan, optimizer_config)?
+            .unwrap_or_else(|| plan.clone()))
```

Comment:
I've stumbled upon this while working on #4614. I think this is wrong: the type coercion MUST NOT be skipped in case of an error, otherwise we may end up with non-executable plans.

Reply:
This PR don't change any process logic.
I don't know about `skipped` you said.

Reply:
I think @crepererum  is saying that if the type coercion rules fail internally (return an Error) it is likely a serious bug in DataFusion and the plan will not work as expected.

By effectively ignoring the error here, the error will appear at some later stage (e.g. can't run the plan), making it harder to debug the source of the issue

🤔

Reply:
Here I don't ignore error.

```rust
            self
            .try_optimize(plan, optimizer_config)?
            .unwrap_or_else(|| plan.clone())
```

error will be return by `?`

`unwrap_or_else` is used for `option<>`

Reply:
OK, let's move the discussion / fix to a follow-up: #4615.